### PR TITLE
#515: Port partition aware policy to C#

### PR DESCRIPTION
**Pull request by spolitov**
Implemented partition aware policy in C# driver.
All primitive and frozen types are supported.
Ported policy tests from Java to C#.

Added JSON support to C# driver.
Ported JSON test from Java to C#.

### CODE REVIEW

File: src/Cassandra/ConsistencyLevel.cs [hunk 1]
```diff
@@ -76,6 +76,19 @@ public enum ConsistencyLevel
         /// <summary>
         /// Similar to <c>One</c> but only within the DC the coordinator is in.
         /// </summary>
-        LocalOne = 0x000A
+        LocalOne = 0x000A,
+        YbConsistentPrefix = One,
```
Comment (robertpang) [inline src/Cassandra/ConsistencyLevel.cs]: Per MS naming convention, 2-letter acronym should be kept in upper case (https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions#capitalization-rules-for-identifiers). Rename to `YBConsistentPrefix` and `YBStrong`.

File: src/Cassandra/ConsistencyLevel.cs [hunk 2]
```diff
@@ -76,6 +76,19 @@ public enum ConsistencyLevel
...
+        YbStrong = Quorum,
```
Comment (robertpang) [inline src/Cassandra/ConsistencyLevel.cs]: Nit: remove trailing comma.

File: src/Cassandra/ConsistencyLevel.cs [hunk 3]
```diff
@@ -76,6 +76,19 @@ public enum ConsistencyLevel
...
+    }
+
+    public static class ConsistencyLevelMethod
+    {
+        /// <summary>
+        /// Whether or not this consistency level corresponds to YB strong consistency.
+        /// </summary> 
+        public static bool IsStrong(this ConsistencyLevel self)
```
Comment (robertpang) [inline src/Cassandra/ConsistencyLevel.cs]: Rename `self` to `level`. Also, unsure about C# syntax. Is `this` needed?

File: src/Cassandra/Metadata.cs [hunk 4]
```diff
@@ -479,5 +485,74 @@ internal void SetCassandraVersion(Version version)
...
+        /// <summary>
+        /// Refresh the partition map for all tables.
+        /// </summary>
+        internal async Task RefreshPartitionMap()
+        {
+            if (!Configuration.Policies.LoadBalancingPolicy.RequiresPartitionMap)
+            {
+                return;
+            }
+            var address2hosts = new Dictionary<IPAddress, Host>();
```
Comment (robertpang) [inline src/Cassandra/Metadata.cs]: Nit: `addressToHosts`?

File: src/Cassandra/Serialization/DataTypeParser.cs [hunk 5]
```diff
@@ -80,7 +80,8 @@ private static class CqlNames
             {"bigint", ColumnTypeCode.Bigint},
             {"decimal", ColumnTypeCode.Decimal},
             {"varint", ColumnTypeCode.Varint},
-            {"counter", ColumnTypeCode.Counter}
+            {"counter", ColumnTypeCode.Counter},
+            {"jsonb", ColumnTypeCode.Json},
```
Comment (robertpang) [inline src/Cassandra/Serialization/DataTypeParser.cs:99]: Nit: remove trailing comma.

File: src/Cassandra/Serialization/DictionarySerializer.cs [hunk 6]
```diff
@@ -34,7 +36,15 @@ public override IDictionary Deserialize(ushort protocolVersion, byte[] buffer, i
...
-            var result = (IDictionary)Activator.CreateInstance(dicType);
+            IDictionary result;
+            object comparer = null;
+            // system.partitions contains dictonary with IPAddress key, that is not comparable.
+            // So use custom comparer for it.
+            if (keyType == typeof(IPAddress))
+            {
+                comparer = IPAddressComparer.Instance;
+            }
+            result = (IDictionary)(comparer == null ? Activator.CreateInstance(dicType) : Activator.CreateInstance(dicType, comparer));
```
Comment (robertpang) [inline src/Cassandra/Serialization/DictionarySerializer.cs]: Will it be simpler and more readable like this?

```
            IDictionary result;
            // system.partitions contains dictonary with IPAddress key, that is not comparable.
            // So use custom comparer for it.
            if (keyType == typeof(IPAddress))
            {
                result = Activator.CreateInstance(dicType, IPAddressComparer.Instance));
            }
            else
            {
                result = Activator.CreateInstance(dicType);
            }
```

File: src/Cassandra/Serialization/Serializer.cs [hunk 7]
```diff
@@ -53,7 +53,8 @@ internal class Serializer
             { ColumnTypeCode.TinyInt, TypeSerializer.PrimitiveSbyteSerializer },
             { ColumnTypeCode.Uuid, TypeSerializer.PrimitiveGuidSerializer },
             { ColumnTypeCode.Varchar, TypeSerializer.PrimitiveStringSerializer },
-            { ColumnTypeCode.Varint, TypeSerializer.PrimitiveBigIntegerSerializer }
+            { ColumnTypeCode.Varint, TypeSerializer.PrimitiveBigIntegerSerializer },
+            { ColumnTypeCode.Json, TypeSerializer.PrimitiveJsonSerializer },
```
Comment (robertpang) [inline src/Cassandra/Serialization/Serializer.cs:71]: Nit: remove trailing comma.

File: src/Cassandra/ConsistencyLevel.cs (hunk 3 above)
Comment (spolitov) [reply src/Cassandra/ConsistencyLevel.cs]: It is extension method.
So one could use it in the following manner:
```
ConsistencyLevel level = ...
if (level.IsStrong()) { ... }
```

File: src/Cassandra/ConsistencyLevel.cs (hunk 2 above)
Comment (spolitov) [reply src/Cassandra/ConsistencyLevel.cs]: Trailing comma allows to add entries w/o modifying last line.
So it is better to review.

File: src/Cassandra/YugaByte/PartitionAwarePolicy.cs [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+using System.Linq;
+using System.Net;
+using System.Text;
+using Cassandra.Serialization;
+
+namespace Cassandra.YugaByte
+{
+    public class PartitionAwarePolicy : ILoadBalancingPolicy
+    {
+        private readonly ILoadBalancingPolicy _childPolicy = new DCAwareRoundRobinPolicy(null, int.MaxValue);
```
Comment (robertpang) [inline src/Cassandra/YugaByte/PartitionAwarePolicy.cs]: This should not be hard-coded but be passed in in the constructor. See `PartitionAwarePolicy.java` counterpart and `TokenAwarePolicy.cs`.

File: src/Cassandra/YugaByte/PartitionMetadata.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        // The partition start -- inclusive bound.
+        public int StartKey { get; private set; }
+        // The partition end -- exclusive bound.
+        public int EndKey { get; private set; }
+        // The list of hosts -- first one should be the leader, then the followers in no particular order.
+        // TODO We should make the leader explicit here and in the return type of the getQueryPlan method
+        // to be able to distinguish the case where the leader is missing so the hosts are all followers.
+        public IEnumerable<Host> Hosts { get; private set; }
+
+        public bool Valid {
```
Comment (robertpang) [inline src/Cassandra/YugaByte/PartitionMetadata.cs]: Do we have a real use case when startKey is < 0? If not, let's remove this attribute.

**Review by robertpang — CHANGES_REQUESTED** [review]
Looks good to me. Should be able to approve the merge after comments are addressed.

File: src/Cassandra/Serialization/DataTypeParser.cs (hunk 5 above)
Comment (spolitov) [reply src/Cassandra/Serialization/DataTypeParser.cs:99]: Trailing comma allows to add entries w/o modifying last line.
So it is better for review.

File: src/Cassandra/Serialization/Serializer.cs (hunk 7 above)
Comment (spolitov) [reply src/Cassandra/Serialization/Serializer.cs:71]: Trailing comma allows to add entries w/o modifying last line.
So it is better for review.

File: src/Cassandra/Policies/Policies.cs [hunk 10]
```diff
@@ -31,7 +33,7 @@ public static ILoadBalancingPolicy DefaultLoadBalancingPolicy
         {
             get
             {
-                return new TokenAwarePolicy(new DCAwareRoundRobinPolicy());
+                return new PartitionAwarePolicy();
```
Comment (robertpang) [inline src/Cassandra/Policies/Policies.cs]: Let's create `PartitionAwarePolicy` by passing the `DCAwareRoundRobinPolicy()` explicitly here like how it was with `TokenAwarePolicy()`. The API doc above needs to be updated also.

File: src/Cassandra/YugaByte/PartitionAwarePolicy.cs [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+using Cassandra.Serialization;
+
+namespace Cassandra.YugaByte
+{
+    public class PartitionAwarePolicy : ILoadBalancingPolicy
+    {
+        private readonly ILoadBalancingPolicy _childPolicy;
+        private ICluster _cluster;
+
+        public PartitionAwarePolicy() : this(new DCAwareRoundRobinPolicy(null, int.MaxValue))
```
Comment (robertpang) [inline src/Cassandra/YugaByte/PartitionAwarePolicy.cs]: Let's remove this default constructor like `TokenAwarePolicy()` does not have one.

**Review by robertpang — APPROVED** [review]
Looks good. Left 2 more comments but is good to merge after addressing.

**Review by rao-vasireddy — APPROVED** [review]
Copyright notice changes look good.

**Pull request merged**